idgen/idgen.go

```diff
@@ -139,7 +145,13 @@ func (bg *BalancedIdentityGenerator) Depth() int {
 }
 
 func (bg *BalancedIdentityGenerator) genID() (crypto.PrivKey, TrieKey, error) {
-	privKey, _, err := crypto.GenerateKeyPairWithReader(crypto.Ed25519, 0, bg.reader)
+	hash := sha256.New
+	info := []byte("hydra keys")
+	seed := bg.seed
+	salt := atomic.AddUint32(&bg.idgenCount, 1)
+	salt_bytes := make([]byte, 4)
```

Comment:
Should the salt here be something like "Hydra-X|Head-Y" instead? It seems like it would be more efficient in that Hydra X+1 doesn't have to run through all the Hydra X keys.

Also, IIUC the offsets are 100 per hydra but each head consumes at least 2 keys since AddBalanced generates two uniqueIDs. Going the offset way the it seems like we'd need at least 200 per hydra unless I'm missing something.

Reply:
All the hydras have to run through the previous keys or they wont make the lookup table for them and therefore won't be deterministic

The offset is applied to AddBalanced, not genID so it's to the used IDs (that wind up in the table, not the generated ones). I thought the tests exercised this pretty well.

I'll add a longish comment with the offset strategy, Petar and I chatted some in slack about it and it's still fresh in my mind

Reply:
Worth saying it's not just that they are deterministic (your strategy would be), but that they remain balanced

Reply:
Ah, yep because we're not coordinating anymore everyone needs to generate the whole sequence themselves.

Reply:
> The offset is applied to AddBalanced, not genID so it's to the used IDs (that wind up in the table) not the generated ones. I thought the tests exercised this pretty well.

Makes sense. I probably should've been more careful when reading, but also adding some comments in the code here should clarify what's going to future readers.